src/main/java/ca/uhn/fhir/jpa/starter/CustomConsentService.java

```diff
@@ -0,0 +1,173 @@
+      return ConsentOutcome.AUTHORIZED;
+    }
+    String patientId = getPatientFromToken(theRequestDetails);
+    String resourceName = null;
+    IBaseResource existingResource = null;
+    boolean proceed = false;
+    switch (theRequestDetails.getRequestType().toString()) {
+      case "POST":
+        resourceName = theRequestDetails.getResourceName();
+        proceed = isResourceValid(resourceName, patientId, theRequestDetails.getResource());
+        break;
+      case "PUT":
+        resourceName = theRequestDetails.getResourceName();
+        existingResource = getResourceFromDB(theRequestDetails.getRequestPath());
+        if (isResourceValid(resourceName, patientId, existingResource)) {
+          proceed = isResourceValid(resourceName, patientId, theRequestDetails.getResource());
+        }
+        break;
+      case "PATCH":
+        resourceName = theRequestDetails.getResourceName();
+        existingResource = getResourceFromDB(theRequestDetails.getRequestPath());
+        if (isResourceValid(resourceName, patientId, existingResource)) {
+          // As Patch request body doesn't contain any Resource we need to handle it
+          // differently
+          proceed = isPatchRequestBodyValid(resourceName,
+              new String(theRequestDetails.getRequestContentsIfLoaded()));
+        }
+        break;
+      default:
+        proceed = true;
+        break;
+    }
+    return proceed ? ConsentOutcome.PROCEED : ConsentOutcome.REJECT;
+  }
+
+  @Override
+  public ConsentOutcome canSeeResource(RequestDetails theRequestDetails, IBaseResource theResource,
+      IConsentContextServices theContextServices) {
+    /*
+     * Returning authorized if there is no Authorization header present or if requested resource is
+     * present in Patient Compartment For both this cases all the consent logic is in authorization
+     * intercepter rules
+     */
+    if (ObjectUtils.isEmpty(theRequestDetails.getHeader("Authorization"))
+        || oAuth2Helper.canBeInPatientCompartment(theRequestDetails.getResourceName())) {
+      return ConsentOutcome.AUTHORIZED;
+    }
+    String patientId = getPatientFromToken(theRequestDetails);
+    String resourceName = theResource.getClass().getSimpleName();
+    return isResourceValid(resourceName, patientId, theResource) ? ConsentOutcome.PROCEED
+        : ConsentOutcome.REJECT;
+  }
+
+  private IBaseResource getResourceFromDB(String requestedPath) {
+    String[] requestDetail = requestedPath.split("/");
+    return daoRegistry.getResourceDao(requestDetail[0]).read(new IdType(requestDetail[1]));
+  }
+
+  private boolean isResourceValid(String resourceName, String patientId,
+      IBaseResource theResource) {
+    if (patientId != null) {
+      String patientRef = "Patient/" + patientId;
+      switch (resourceName) {
+        case "Task":
+          return isReferanceValid((Resource) theResource, patientRef, "for");
```

Comment:
Did you try this with some example data? The fhir specs have a few different ways to "reference" a patient from a task.

1. `for`: The is documented as "Beneficiary of the Task". Seems applicable.
2. `requester`: This is documented as "Who is asking for task to be done". This doesn't look like the right reference for this use case.
3. `owner`: This is documented as "Responsible individual". I'm not sure what that means.
4. `restriction` `recipient`: This is documented as "For whom is fulfillment sought?". That seems like it might also be applicable.

@boxwala, I did a little more research but don't see documentation on which should be used for this. Is `for` the right reference?